src/main/java/com/minecolonies/coremod/compatibility/jei/transer/PrivateCraftingTeachingTransferHandler.java

```diff
@@ -0,0 +1,69 @@
+package com.minecolonies.coremod.compatibility.jei.transer;
+
+import com.google.common.collect.ImmutableSet;
+import com.minecolonies.api.util.ItemStackUtils;
+import com.minecolonies.coremod.MineColonies;
+import com.minecolonies.coremod.inventory.CraftingGUIBuilding;
+import com.minecolonies.coremod.network.messages.TransferRecipeCrafingTeachingMessage;
+import mezz.jei.api.gui.IGuiIngredient;
+import mezz.jei.api.gui.IGuiItemStackGroup;
+import mezz.jei.api.gui.IRecipeLayout;
+import mezz.jei.api.recipe.transfer.IRecipeTransferError;
+import mezz.jei.api.recipe.transfer.IRecipeTransferHandler;
+import mezz.jei.api.recipe.transfer.IRecipeTransferHandlerHelper;
+import net.minecraft.entity.player.EntityPlayer;
+import net.minecraft.inventory.Slot;
+import net.minecraft.item.ItemStack;
+import net.minecraft.util.Tuple;
+import net.minecraft.util.text.translation.I18n;
+
+import javax.annotation.Nullable;
+import java.util.*;
+import java.util.stream.Collectors;
+
+public class PrivateCraftingTeachingTransferHandler implements IRecipeTransferHandler<CraftingGUIBuilding>
+{
+    private final IRecipeTransferHandlerHelper handlerHelper;
+
+    public PrivateCraftingTeachingTransferHandler(final IRecipeTransferHandlerHelper handlerHelper) {this.handlerHelper = handlerHelper;}
+
+    @Override
+    public Class<CraftingGUIBuilding> getContainerClass()
+    {
+        return CraftingGUIBuilding.class;
+    }
+
+    @Nullable
+    @Override
+    public IRecipeTransferError transferRecipe(
+      final CraftingGUIBuilding craftingGUIBuilding, final IRecipeLayout recipeLayout, final EntityPlayer entityPlayer, final boolean b, final boolean b1)
+    {
+        IGuiItemStackGroup itemStackGroup = recipeLayout.getItemStacks();
+
+        // indexes that do not fit into the player crafting grid
+        Set<Integer> badIndexes = ImmutableSet.of(2, 5, 6, 7, 8);
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Assign this magic number 5 to a well-named constant, and use the constant instead. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Assign this magic number 6 to a well-named constant, and use the constant instead. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Assign this magic number 7 to a well-named constant, and use the constant instead. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Assign this magic number 8 to a well-named constant, and use the constant instead. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)